webapp/src/lib/encryption.ts

```diff
@@ -0,0 +1,194 @@
+	);
+	const privateKey = await crypto.subtle.importKey(
+		"pkcs8",
+		pkcs8,
+		{ name: "X25519" },
+		false,
+		["deriveBits"],
+	);
+
+	// Unfortunately, Crypto Subtle API does not have a good way to derive an
+	// X25519 public key from a private one - exporting as `jwk` from a PKCS #8
+	// imported private key does not work across browsers. Use `@noble/curves`
+	// to compute the raw public key instead.
+	const publicKeyRaw = x25519.getPublicKey(ethers.getBytes(secret));
+	const publicKey = await importPublicKey(publicKeyRaw);
+
+	return { publicKey, privateKey };
+}
+
+/**
+ * Import a raw x25519 public key to a Crypto Subtle key.
+ */
+async function importPublicKey(raw: BytesLike): Promise<CryptoKey> {
+	if (ethers.dataLength(raw) !== 32) {
+		throw new Error("invalid raw x25519 public key");
+	}
+	return await crypto.subtle.importKey(
+		"raw",
+		ethers.getBytes(raw),
+		{ name: "X25519" },
+		true,
+		[],
+	);
+}
+
+/**
+ * Exports a public key to its raw bytes.
+ */
+async function exportPublicKey({
+	publicKey,
+}: Pick<CryptoKeyPair, "publicKey">): Promise<Hex> {
+	const raw = await crypto.subtle.exportKey("raw", publicKey);
+	return ethers.hexlify(new Uint8Array(raw));
+}
+
+/**
+ * RLP-encode a Safe transaction for encryption.
+ */
+function rlpEncodeSafeTransaction(safeTx: SafeTransaction): Uint8Array {
+	const n = ethers.toBeArray;
+	return ethers.getBytes(
+		ethers.encodeRlp([
+			safeTx.to,
+			n(safeTx.value),
+			safeTx.data,
+			n(safeTx.operation),
+			n(safeTx.safeTxGas),
+			n(safeTx.baseGas),
+			n(safeTx.gasPrice),
+			safeTx.gasToken,
+			safeTx.refundReceiver,
+		]),
+	);
+}
+
+/**
+ * RLP-decode a Safe transaction for decryption.
+ */
+function rlpDecodeSafeTransaction(data: BytesLike): SafeTransaction {
+	const decoded = ethers.decodeRlp(data);
+	if (
+		!Array.isArray(decoded) ||
+		decoded.length !== 9 ||
+		decoded.some((field) => typeof field !== "string")
+	) {
+		throw new Error("invalid Safe transaction RLP encoding");
+	}
+	const fields = decoded as string[];
+
+	const a = ethers.getAddress;
+	const n = (field: string) =>
+		field === "0x"
+			? "0x0"
+			: ethers.toQuantity(BigInt(ethers.toBeHex(field, 32)));
+	const o = (field: string) => {
+		const value = n(field);
+		if (value === "0x0") {
+			return 0;
+		}
+		if (value === "0x1") {
+			return 1;
+		}
+		throw new Error(`invalid Safe operation ${BigInt(field)}`);
+	};
+
+	return {
+		to: a(fields[0]),
+		value: n(fields[1]),
+		data: fields[2],
+		operation: o(fields[3]),
+		safeTxGas: n(fields[4]),
+		baseGas: n(fields[5]),
+		gasPrice: n(fields[6]),
+		gasToken: a(fields[7]),
+		refundReceiver: a(fields[8]),
+	};
+}
+
+/**
+ * Encrypt a Safe transaction for submission to Secret Harbour.
+ */
+async function encryptSafeTransaction(
+	transaction: SafeTransaction,
+	{ privateKey }: Pick<CryptoKeyPair, "privateKey">,
+	recipientPublicKeys: CryptoKey[],
+): Promise<Hex> {
+	// TODO: `jose` generates a random x25519 private key per receiver for
```

Comment:
tbh I am not sure what the TODO is here right now 😅 

I assume it is to use this passed private key instead of using the random private key generated by jose?

Reply:
Yes.